components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -146,6 +155,16 @@ class GeckoEngine(
         return GeckoEngineSessionState.fromJSON(json)
     }
 
+    /**
+     * See [Engine.speculativeCreateSession].
+     */
+    override fun speculativeCreateSession(private: Boolean) {
+        val speculativeSession = this.speculativeSession
+        if (speculativeSession == null || speculativeSession.geckoSession.settings.usePrivateMode != private) {
+            this.speculativeSession = GeckoEngineSession(runtime, private, defaultSettings)
```

Comment:
In the case where `usePrivateMode != private` we drop the previous `speculativeSession`. Internally `GeckoEngineSession` calls `open()` on the `GeckoSession`. Wouldn't we need to call `close()` here to avoid leaking that session since we never told Gecko to close it?

Reply:
Related to that I wonder if, since both methods are not syncrhonized, we could end up in a situation where we create two sessions in a race and one gets leaked and never closed? Same concern for `createSession()`: Could we end up in a situation where we, due to a race, end up using the same `GeckoEngineSession` twice? Of course all that depends on the calling thread - if it is always `main` then I guess it doesn't matter.

Reply:
> Wouldn't we need to call close() here to avoid leaking that session since we never told Gecko to close it?

Ah, yes, good catch! I'll update in a bit.

> ... if it is always main then I guess it doesn't matter.

Yes, we really want to be on the main thread for these methods. Let's add an assert on main thread for `speculativeCreateSession` to prevent weirdness. This way we can also get rid of `volatile` which wouldn't be sufficient.

Reply:
OK, updated.